### PR TITLE
introduce multiple private states and multi-tenancy support

### PR DESCRIPTION
**Pull request by trung**
### Summary

Multiple Private States is a feature that allows a GoQuorum node to manage more than one private state. The private states are separated logically in the shared database.

This functionality lays the foundation for supporting multiple tenants on a single node. 

**The current multi-tenancy feature will be replaced by this PR**

### Performance - Update from Nam:

PASSED: Performance tests have been executed to determine the impact of having multiple private states and results demonstrated that the difference aren’t significant; moreover, the increase in the number of tenants in a single node does not cause a drop in the network’s performance.

### Note

* `import`/`export` commands are not supported in this PR

### CODE REVIEW

File: consensus/clique/snapshot_test.go [hunk 1]
```diff
@@ -450,7 +450,7 @@ func TestClique(t *testing.T) {
 		// Pass all the headers through clique and ensure tallying succeeds
 		chain, err := core.NewBlockChain(db, nil, &config, engine, vm.Config{}, nil, nil)
 		if err != nil {
-			t.Errorf("test %d: failed to create test chain: %v", i, err)
+			t.Errorf("test %d`: failed to create test chain: %v", i, err)
```
Comment (ricardolyn) [inline consensus/clique/snapshot_test.go]: mistake?

File: core/blockchain.go [hunk 2]
```diff
@@ -666,22 +693,52 @@ func (bc *BlockChain) Processor() Processor {
...
+func (bc *BlockChain) State() (*state.StateDB, mps.PrivateStateRepository, error) {
 	return bc.StateAt(bc.CurrentBlock().Root())
 }
 
-// StateAt returns a new mutable state based on a particular point in time.
-func (bc *BlockChain) StateAt(root common.Hash) (*state.StateDB, *state.StateDB, error) {
-	publicStateDb, privateStateDb, err := state.NewDual(root, bc.stateCache, bc.snaps, bc.db, bc.privateStateCache, nil)
+// StatePSI returns a new mutable public state and a mutable private state for given PSI,
+// based on the current HEAD block.
+func (bc *BlockChain) StatePSI(psi types.PrivateStateIdentifier) (*state.StateDB, *state.StateDB, error) {
```
Comment (ricardolyn) [inline core/blockchain.go:704]: I think is always better to add Quorum functions at the end of the file as it simplifies geth upgrades

File: core/state/statedb.go [hunk 3]
```diff
@@ -159,6 +159,7 @@ func New(root common.Hash, db Database, snaps *snapshot.Tree) (*StateDB, error)
 
 // Quorum
 // NewDual - Create a public and private state from a given public and private tree
+// TODO - rebase - this is not used (except in tests). It was used in the trace APIs. Should we remove it?
```
Comment (ricardolyn) [inline core/state/statedb.go]: if not used, then we should remove it. should not be left only for tests

File: miner/worker.go [hunk 4]
```diff
@@ -785,25 +786,46 @@ func (w *worker) updateSnapshot() {
 }
 
 func (w *worker) commitTransaction(tx *types.Transaction, coinbase common.Address) ([]*types.Log, error) {
-	snap := w.current.state.Snapshot()
-	privateSnap := w.current.privateState.Snapshot()
-
+	workerEnv := w.current
```
Comment (ricardolyn) [inline miner/worker.go:789]: would make sense to encapsulate this block of code before the `core.ApplyTransaction(..)` into a different quorum specific function?

File: rpc/context.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
+package rpc
```
Comment (ricardolyn) [inline rpc/context.go:2]: could we add some notes at the top of the file or a different name of the file to know that this is a Quorum specific file?

**Review by ricardolyn — COMMENTED** [review]
Minor comments. Some functions might be missing the Quorum comments

File: consensus/clique/snapshot_test.go (hunk 1 above)
Comment (trung) [reply consensus/clique/snapshot_test.go]: fixed

File: core/blockchain.go (hunk 2 above)
Comment (trung) [reply core/blockchain.go:704]: added

File: core/state/statedb.go (hunk 3 above)
Comment (trung) [reply core/state/statedb.go]: removed

File: rpc/context.go (hunk 5 above)
Comment (trung) [reply rpc/context.go:2]: added

**Review by ricardolyn — APPROVED** [review]
LGTM 👍

**Pull request merged**